packages/backend-common/src/reading/tree/ArchiveResponse.ts

```diff
@@ -0,0 +1,160 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import tar, { Parse, ParseStream, ReadEntry } from 'tar';
+import path from 'path';
+import fs from 'fs-extra';
+import { Readable, pipeline as pipelineCb } from 'stream';
+import { promisify } from 'util';
+import concatStream from 'concat-stream';
+import {
+  ReadTreeResponse,
+  ReadTreeResponseFile,
+  ReadTreeResponseDirOptions,
+} from '../types';
+
+// Tar types for `Parse` is not a proper constructor, but it should be
+const TarParseStream = (Parse as unknown) as { new (): ParseStream };
+
+const pipeline = promisify(pipelineCb);
+
+/**
+ * Wraps a tar archive stream into a tree response reader.
+ */
+export class ArchiveResponse implements ReadTreeResponse {
+  private read = false;
+
+  constructor(
+    private readonly stream: Readable,
+    private readonly subPath: string,
+    private readonly workDir: string,
+    private readonly filter?: (path: string) => boolean,
+  ) {
+    if (subPath) {
+      if (!subPath.endsWith('/')) {
+        this.subPath += '/';
+      }
+      if (subPath.startsWith('/')) {
+        throw new TypeError(
+          `ArchiveResponse subPath must not start with a /, got '${subPath}'`,
+        );
+      }
+    }
+  }
+
+  // Make sure the input stream is only read once
+  private onlyOnce() {
+    if (this.read) {
+      throw new Error('Response has already been read');
+    }
+    this.read = true;
+  }
+
+  async files(): Promise<ReadTreeResponseFile[]> {
+    this.onlyOnce();
+
+    const files = Array<ReadTreeResponseFile>();
+    const parser = new TarParseStream();
+
+    parser.on('entry', (entry: ReadEntry & Readable) => {
+      if (entry.type === 'Directory') {
+        entry.resume();
+        return;
+      }
+
+      if (this.subPath) {
+        if (!entry.path.startsWith(this.subPath)) {
+          entry.resume();
+          return;
+        }
+      }
+
+      const path = entry.path.slice(this.subPath.length);
+      if (this.filter) {
+        if (!this.filter(path)) {
+          entry.resume();
+          return;
+        }
+      }
+
+      const content = new Promise<Buffer>(async resolve => {
+        await pipeline(entry, concatStream(resolve));
+      });
+
+      files.push({ path, content: () => content });
+
+      entry.resume();
+    });
+
+    await pipeline(this.stream, parser);
+
+    return files;
+  }
+
+  async archive(): Promise<Readable> {
+    if (!this.subPath) {
+      this.onlyOnce();
+
+      return this.stream;
+    }
+
+    // TODO(Rugvip): method for repacking a tar with a subpath is to simply extract into a
+    //               tmp dir and recreate the archive. Would be nicer to stream things instead.
```

Comment:
I banged my head against trying to do that for a day or so until i gave up and just left it. I'm sure it can be done, probably using tar.Pack. but the documentation of the tar lib is not easy to understand unless you already know how it all works. :/